core/amber/src/main/scala/edu/uci/ics/texera/web/service/ResultExportService.scala

```diff
@@ -362,4 +337,153 @@ class ResultExportService(opResultStorage: OpResultStorage, wId: UInteger) {
+      channel: WritableByteChannel
+  ): (ArrowFileWriter, VectorSchemaRoot) = {
+    val schema = results.head.getSchema
+    val arrowSchema = ArrowUtils.fromTexeraSchema(schema)
+    val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    val writer = new ArrowFileWriter(root, null, channel)
+    (writer, root)
+  }
+
+  private def writeArrowData(
+      writer: ArrowFileWriter,
+      root: VectorSchemaRoot,
+      results: Iterable[Tuple]
+  ): Unit = {
+    writer.start()
+    val batchSize = 1000
+
+    // Convert to Seq to get total size
+    val resultSeq = results.toSeq
+    val totalSize = resultSeq.size
+
+    // Process in complete batches
+    for (batchStart <- 0 until totalSize by batchSize) {
+      val batchEnd = Math.min(batchStart + batchSize, totalSize)
+      val currentBatchSize = batchEnd - batchStart
+
+      // Process each tuple in the current batch
+      for (i <- 0 until currentBatchSize) {
+        val tuple = resultSeq(batchStart + i)
+        ArrowUtils.setTexeraTuple(tuple, i, root)
+      }
+
+      // Set the correct row count for this batch and write it
+      root.setRowCount(currentBatchSize)
+      writer.writeBatch()
+      root.clear()
+    }
+
+    writer.end()
+  }
+
+  private def finalizeArrowExport(
+      request: ResultExportRequest,
+      user: User,
+      stream: ByteArrayOutputStream,
+      extension: String
+  ): ResultExportResponse = {
+    val fileName = generateFileName(request, extension)
+    saveToDatasets(request, user, stream.toByteArray, fileName)
+    ResultExportResponse(
+      "success",
+      s"Arrow file saved as $fileName to Datasets ${request.datasetIds.mkString(",")}"
+    )
+  }
+
+  private def generateFileName(request: ResultExportRequest, extension: String): String = {
+    val latestVersion =
+      WorkflowVersionResource.getLatestVersion(UInteger.valueOf(request.workflowId))
+    val timestamp = LocalDateTime
+      .now()
+      .truncatedTo(ChronoUnit.SECONDS)
+      .format(DateTimeFormatter.ofPattern("yyyy-MM-dd_HH-mm-ss"))
+    sanitizePath(
+      s"${request.workflowName}-v$latestVersion-${request.operatorName}-$timestamp.$extension"
+    )
+  }
+
+  private def saveToDatasets(
+      request: ResultExportRequest,
+      user: User,
+      data: Array[Byte],
+      fileName: String
+  ): Unit = {
+    val fileStream = new ByteArrayInputStream(data)
+    request.datasetIds.foreach { did =>
+      val datasetPath = PathUtils.getDatasetPath(UInteger.valueOf(did))
+      val filePath = datasetPath.resolve(fileName)
+      createNewDatasetVersionByAddingFiles(
+        UInteger.valueOf(did),
+        user,
+        Map(filePath -> fileStream)
+      )
+    }
+  }
+
+  private def convertFieldToBytes(field: Any): Array[Byte] = {
+    field match {
+      case data: Array[Byte] => data
+      case data: String      => data.getBytes(StandardCharsets.UTF_8)
+      case data              => data.toString.getBytes(StandardCharsets.UTF_8)
```

Comment:
Thank you for pointing this out. I reviewed the behavior and found that the CSV scan operator converts any CSV file (regardless of its encoding) into a Texera table in UTF-8 format during the conversion process. I tested this by uploading a UTF-16 encoded CSV file, scanning it with the CSV scan operator, and then downloading the result. When I tried using the resulting file in another workflow with the CSV scan operator and specified UTF-16 encoding, it failed to read the file. However, when using the UTF-8 option, it worked as expected.

Based on this behavior, the current implementation of Arrow serialization and deserialization assumes UTF-8 encoding. Allowing users to choose an encoding scheme might introduce inconsistencies since Workflow 1 always outputs data in UTF-8.

Reply:
correct. we had the assumption to use utf8 encoding, mainly for user-facing serializations (i.e., user need to be able to read the file). However, as we are talking about a more general serialization format, and it is not user-facing (i.e., we don't expect users to open and read an arrow format file), but rather a machine-facing format to support any arbitrary bytes, I wonder if hard code to utf8 still makes sense. 

There are three questions here:
1. Do we have to specify one encoding for binary data? as discussed before, it should be better to save the bytes "as it is". If we hardcode it to utf8, then I could also achieve the same effect by using "base64" in CSV format.
2. If we have to specify one encoding, is utf8 good enough? for example if the bytes are image, music, or video data, converting to utf8 does it lose any precision? 
3. If we have to have one encoding, does it make sense to let the user specify the encoding?

Reply:
I believe there may be some misunderstandings regarding the scope of this PR. Specifically, the code you referenced is not directly related to the core functionality being addressed here—it is part of the serialization logic for saving individual cells into a file. If necessary, I can remove this change from the PR as it is primarily a refactoring effort.

Regarding the handling of binary data, it’s important to note that binary data is written directly to the `VarBinaryVector` (see `ArrowUtils.scala`). This ensures that the exact byte representation of the binary data is preserved without any modification. Consequently, the encoding concerns you raised do not apply in this context, as the data is stored in its original binary format.

I hope this addresses your concerns.

Reply:
you are saying those code block is used by export as CSV, but not used by the Arrow format?

Reply:
No, it is only used to export a cell, not a table.

Reply:
ok